File: find_minimum_in_rotated_sorted_array/Solutions.py
class Solution(object):
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return None
        if len(nums) == 1:
            return nums[0]

        left = 0
        right = len(nums)-1

        minimum = nums[0]

        while left < right:
            # print(str(left) + " -> " + str(right))
            mid = (left + right) >> 1
            if nums[left] < nums[mid]: # the left side is sorted. So, the min element in right side
                left = mid
            if nums[left] > nums[mid]:
                right = mid
            if right == left + 1: # In case of an array doesn't have a pivot
                return min(nums[right],minimum)


if __name__ == "__main__":
    nums = [1,2]
    test = Solution()
    print(test.findMin(nums))

File: minimum_size_subarray_sum/Solutions.cc
#include <iostream>
#include <vector>

using std::cin;
using std::cout;
using std::endl;

using std::vector;

class Solution {
public:
    int minSubArrayLen(int s, vector<int>& nums) {
        int len = nums.size();
        if (!len) {
            return 0;
        }
        
        int left = 0, right = 0, min_length = INT16_MAX, sum = 0;
        while (right < len) {
            while (sum < s && right < len) {
                sum += nums[right++];
            }
            while (sum >= s) {
                int curr_length = right - left;
                min_length = min_length < curr_length ? min_length : curr_length;
                sum -= nums[left++];
            }
        }
        return min_length > len ? 0 : min_length;
    }
};

int main(){
    Solution test;
    vector<int> nums;
    nums.push_back(1);
    nums.push_back(2);
    nums.push_back(3);
    nums.push_back(4);
    nums.push_back(5);
    
    cout << "The minimum size is: " << test.minSubArrayLen(11,nums) << endl;
}

